### PR TITLE
[New feature] Add transformer for language model

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

Add transformer to support language model, including:
. Layer normalization implement
. Prepare decode implement
. Transformer implement (only support LM now)
. Expand dims when handing two tensors with different size
. Optimize backward graph generation to keep topology order
. python api support

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: pyspark/bigdl/nn/layer.py [hunk 1]
```diff
@@ -930,6 +930,21 @@ def __init__(self, hidden_size, filter_size, relu_dropout, bigdl_type="float"):
         super(FeedForwardNetwork, self).__init__(None, bigdl_type,
                                         hidden_size, filter_size, relu_dropout)
 
+class TransformerLayer(Layer):
+
+    '''
+    Implementation TransformerLayer
+    >>> ffn = TransformerLayer(20, 4, 2, 3, 1, 0.1, 0.1, 0.1)
```
Comment (wzhongyuan) [inline pyspark/bigdl/nn/layer.py]: ffn -> transformer ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala [hunk 2]
```diff
@@ -38,6 +41,31 @@ class CAddTable[T: ClassTag, D: ClassTag](val inplace: Boolean = false)(
 
   output = Tensor[D]()
 
+  private def canFastBroadcast(tensor: Tensor[D], other: Tensor[D]): Boolean = {
+    if (tensor.nDimension != other.nDimension()) {
+      return false
+    }
+    var d = other.nDimension()
+    while(d > 0) {
+      if (tensor.size(d) != 1 && tensor.size(d) != other.size(d)) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala]: why we need to check for non-squeezed tensor ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala [hunk 3]
```diff
@@ -38,6 +41,31 @@ class CAddTable[T: ClassTag, D: ClassTag](val inplace: Boolean = false)(
...
+        return false
+      }
+      d -= 1
+    }
+    return true
+  }
+
+  private def sumAlongDims(tensor: Tensor[D], other: Tensor[D]): Tensor[D] = {
+    val size = tensor.size()
+    var target: Tensor[D] = other
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CAddTable.scala]: why define `target` ?  you can use `other` directly

**Comment by wzhongyuan** [comment]
Split PR into several independent ones ?

**Comment by zhangxiaoli73** [comment]
Have splited this PR into several independent ones

**Pull request closed without merging**